### PR TITLE
TEST: Produce JSON logs from pytest runs

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

closes https://github.com/intel/scikit-learn-intelex/pull/2087

This PR modifies the script that runs unit tests to also produce a JSON report of the tests that were executed, using a pytest plugin for this purpose.

The logs are only produced when using the 'run_test' script, and only when receiving argument `--json-report` - it does not modify the commands from .yaml files which might also execute unit tests, and does not add this argument `--json-report` to the .yaml calls to these scripts.

---

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [inline conda-recipe/run_test.bat:50]: JSON report generation is not necessary part for conda recipe building.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [reply conda-recipe/run_test.bat:50]: Thanks for pointing this out. One question in this regard: are these scripts (`run_test.sh` and `run_test.bat`) used by the conda recipe, or do they only run the commands in the `meta.yaml` file?

Comment (Alexsandruss) [reply conda-recipe/run_test.bat:50]: `conda-build` seems to prioritize scripts over recipe commands.

Comment (david-cortes-intel) [reply conda-recipe/run_test.bat]: Modified it to produce these logs only when receiving an argument `--json-report`.

**Comment by icfaust** [comment]
@david-cortes-intel I'll review when we get CI's green

**Comment by david-cortes-intel** [comment]
> @david-cortes-intel I'll review when we get CI's green

Fixed the windows issues - all CI jobs are passing now.

**Comment by david-cortes-intel** [comment]
Made some changes to only add the argument `--json-report-file` when receiving argument `--json-report`, in response to @icfaust 's comment about the argument resulting in an error instead of being ignored when `pytest-json-report` is not installed in the python environment.

File: .gitignore
Comment (Alexsandruss) [inline .gitignore:35]: Is there a reason to hide reports? For example, `_pytest_reports` will be visible

**Comment by ethanglaser** [comment]
I see mention of presence of json-report package in public CI logs but beyond that is there a way to validate this is running as intended?

File: .gitignore
Comment (david-cortes-intel) [reply .gitignore:35]: My reasoning was that: (a) the files in the folder are not part of the package, so one might not want to see them side-by-side with them; (b) typically, reporter tools such as coverage and similar create hidden files by default, such as `.coverage`.

**Comment by david-cortes-intel** [comment]
> I see mention of presence of json-report package in public CI logs but beyond that is there a way to validate this is running as intended?

Modified it to generate the reports as part of the CI runs, but not as part of the conda build. Though the reports are not uploaded anywhere.

**Review by samir-nasibli — COMMENTED** [review]
Thank you @david-cortes-intel 
But the general question for me why json reports even needed?
@icfaust @Alexsandruss please take a final look on the PR

File: .ci/pipeline/build-and-test-lnx.yml
Comment (icfaust) [inline .ci/pipeline/build-and-test-lnx.yml:58]: Since we run the CI tests with the json reports, do we do anything with them? (should be stored in an artifact?)

**Review by icfaust — COMMENTED** [review]
Looking pretty good, I have one final question before approval.

File: .ci/pipeline/build-and-test-lnx.yml
Comment (david-cortes-intel) [reply .ci/pipeline/build-and-test-lnx.yml:58]: No, they don't go anywhere - it was just added to verify that the command works and the files are produced.

The reports might get quite big though (>20mb) so not sure if it's appropriate to upload them somewhere.

Comment (icfaust) [reply .ci/pipeline/build-and-test-lnx.yml:58]: If we generate them but don't use them, then its a test of CI and not of the codebase.  I don't think its necessary to run it on every CI job, unless you plan on removing these lines before merging.  If you really want it in, I would run it in the nightly. https://github.com/intel/scikit-learn-intelex/blob/main/.ci/pipeline/nightly.yml#L99 It will show up on main the next day.

Comment (david-cortes-intel) [reply .ci/pipeline/build-and-test-lnx.yml:58]: Thanks for the suggestion. I see the nightly jobs run the commands from this same file - do you suggest to remove the argument `--json-report` from the command in `.github/workflows/ci.yml` then, or is there some other way to make it triggerable for the nightly job but not for the others?

Comment (icfaust) [reply .ci/pipeline/build-and-test-lnx.yml:58]: I think you may have to set a new azure pipelines variable in ci.yml and nightly.yml that the underlying linux and win yamls will see and trigger from (I hope thats okay).

**Comment by icfaust** [comment]
@Alexsandruss Thoughts on this PR?  Have sufficient changes been made to warrant removing the 'Changes Requested'?

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**
